File: data/figures/chapterD/figure_D_3/plot_D_3.py
#!/usr/bin/env python3
"""Replot this thesis figure from the sibling CSV only."""
from pathlib import Path
import argparse
import csv
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

HERE = Path(__file__).resolve().parent
DATA = HERE / "figure_D_3.csv"
parser = argparse.ArgumentParser(description="Replot thesis Figure D.3 from its sibling CSV.")
parser.add_argument("--output", type=Path, help="Output image path; defaults beside this script.")
arguments = parser.parse_args()
OUTPUT = (arguments.output if arguments.output is not None else HERE / "plot_D_3.png").resolve()
OUTPUT.parent.mkdir(parents=True, exist_ok=True)
plt.rcParams.update({"font.family": "serif", "font.serif": ["Times New Roman", "DejaVu Serif"], "font.size": 9})

def rows_from(path):
    with path.open(newline="", encoding="utf-8-sig") as handle:
        return list(csv.DictReader(handle))

def number(row, key):
    return float(row[key])

rows=rows_from(DATA)
profile=[r for r in rows if r["record_type"]=="profile"]
survival=[r for r in rows if r["record_type"]=="survival"]
profile=sorted(profile,key=lambda r:number(r,"input_physical"))
x=np.array([number(r,"input_physical") for r in profile])
thresholds=np.array(sorted({number(r,"normalized_hvi_threshold") for r in survival}))
xi=np.array(sorted({number(r,"input_physical") for r in survival}))
lookup={(number(r,"normalized_hvi_threshold"),number(r,"input_physical")):r for r in survival}
prob=np.array([[number(lookup[(t,v)],"conditional_exceedance_probability") for v in xi] for t in thresholds])
display=np.array([[number(lookup[(t,v)],"display_log10_exceedance_probability") for v in xi] for t in thresholds])
fig,axes=plt.subplots(3,1,figsize=(7.5,10.2))
ax=axes[0]
mesh=ax.pcolormesh(xi,np.log10(thresholds),display,shading="auto",cmap="viridis",vmin=np.log10(1/2048),vmax=0)
cs=ax.contour(xi,np.log10(thresholds),prob,levels=[.01,.1,.5],colors="#222222",linewidths=.8)
ax.clabel(cs,fontsize=7)
ax.set(xlabel="Aspect ratio, AR [-]",ylabel="log10 normalized HVI threshold",title="A. Conditional HVI survival field")
ax.set_yticks([-8,-6,-4,-2,0], labels=["$10^{-8}$","$10^{-6}$","$10^{-4}$","$10^{-2}$","$10^0$"])
ax.set_ylabel("Normalized HVI threshold [-]")
bar=fig.colorbar(mesh,ax=ax,label="Conditional exceedance probability")
bar.set_ticks([np.log10(1/2048),-3,-2,-1,0], labels=["<=1/2048","$10^{-3}$","$10^{-2}$","$10^{-1}$","1"])
def positive(name):
    a=np.array([number(r,name) for r in profile])
    return np.where(a>0,a,np.nan)
ax=axes[1]
ax.plot(x,positive("p90_normalized_hvi"),color="#D59E1A",lw=1.6,label="90th percentile")
ax.plot(x,positive("p99_normalized_hvi"),color="#E86E17",lw=1.8,label="99th percentile")
ax.plot(x,positive("mean_normalized_hvi"),"--",color="#1769AA",lw=1.7,label="Conditional mean")
ax.plot(x,positive("sampled_profile_maximum_normalized_hvi"),color="#222222",lw=2,label="Sampled profile maximum")
ax.scatter([number(profile[0],"selected_input_physical")],[number(profile[0],"selected_hvi_normalized")],marker="*",s=90,zorder=6,clip_on=False,color="#B33A3A",edgecolors="white",label="Selected design")
ax.set_yscale("log"); ax.set_ylim(1e-8,1.15)
ax.set(xlabel="Aspect ratio, AR [-]",ylabel="Normalized HVI [-]",title="B. Conditional quantiles and profile")
ax.grid(True,which="both",color="#d9d9d9",lw=.5); ax.legend(fontsize=7)
ax=axes[2]
ax.plot(x,positive("positive_hvi_fraction"),color="#E86E17",lw=2,label="Positive-HVI fraction")
ax.plot(x,positive("mean_normalized_hvi"),"--",color="#1769AA",lw=1.8,label="Conditional mean")
ax.set_yscale("log"); ax.set_ylim(1e-8,1)
ax.set(xlabel="Aspect ratio, AR [-]",ylabel="Conditional domain-volume statistic [-]",title="C. Positive-HVI support and conditional mean")
ax.grid(True,which="both",color="#d9d9d9",lw=.5); ax.legend(fontsize=8)
fig.suptitle("Frozen sampled-HVI field versus aspect ratio before evaluation 100",fontweight="bold")
for panel in axes:
    panel.set_xlim(float(xi[0]), float(xi[-1]))
    panel.grid(False, which="minor")
    panel.tick_params(labelsize=9)
positive_indices=np.flatnonzero(np.array([number(r,"positive_hvi_fraction") for r in profile])>0)
if len(positive_indices) and positive_indices[-1]+1 < len(profile):
    zero_start=number(profile[positive_indices[-1]+1],"input_physical")
    axes[1].axvline(zero_start, color="#777777", linestyle=":", linewidth=.8)
    axes[1].text(.98,.20,f"All sampled HVI = 0 for AR >= {zero_start:g}",
                 transform=axes[1].transAxes, ha="right", va="center", fontsize=8,
                 bbox=dict(facecolor="white", edgecolor="none", alpha=.9))
fig.tight_layout(pad=1.0,h_pad=2.0); fig.savefig(OUTPUT,dpi=300,bbox_inches="tight"); plt.close(fig); print(OUTPUT)

File: data/figures/chapterD/figure_D_3/plot_D_3.m
function outputFile = plot_D_3(outputFile)
scriptDir = fileparts(mfilename('fullpath'));
if nargin < 1 || isempty(outputFile)
    outputFile = fullfile(scriptDir, 'plot_D_3.png');
end
outputFile = char(outputFile);
outputDir = fileparts(outputFile);
if ~isempty(outputDir) && ~isfolder(outputDir), mkdir(outputDir); end
dataFile = fullfile(scriptDir, 'figure_D_3.csv');
T = readtable(dataFile, 'TextType', 'string', 'VariableNamingRule', 'preserve');
set(groot, 'defaultAxesFontName', 'Times New Roman'); set(groot, 'defaultTextFontName', 'Times New Roman');
P = sortrows(T(T.record_type == "profile", :), 'input_physical'); S = T(T.record_type == "survival", :);
x = unique(S.input_physical,'sorted'); threshold = unique(S.normalized_hvi_threshold,'sorted');
probability = reshape(S.conditional_exceedance_probability,numel(threshold),numel(x));
displayValue = reshape(S.display_log10_exceedance_probability,numel(threshold),numel(x));
fig=figure('Color','w','Visible','off','Position',[100 100 850 1050]); tl=tiledlayout(fig,3,1,'TileSpacing','compact','Padding','compact');
ax=nexttile(tl); imagesc(ax,x,log10(threshold),displayValue); set(ax,'YDir','normal'); hold(ax,'on'); [c,h]=contour(ax,x,log10(threshold),probability,[.01 .1 .5],'LineColor',[.13 .13 .13],'LineWidth',.8); clabel(c,h,'FontSize',7); xlabel(ax,'Aspect ratio, AR [-]'); ylabel(ax,'Normalized HVI threshold [-]'); yticks(ax,[-8 -6 -4 -2 0]); yticklabels(ax,{'10^{-8}','10^{-6}','10^{-4}','10^{-2}','10^{0}'}); title(ax,'A. Conditional HVI survival field'); cb=colorbar(ax); ylabel(cb,'Conditional exceedance probability'); cb.Ticks=[log10(1/2048) -3 -2 -1 0]; cb.TickLabels={'<=1/2048','10^{-3}','10^{-2}','10^{-1}','1'}; clim(ax,[log10(1/2048) 0]);
ax=nexttile(tl); hold(ax,'on'); semilogy(ax,P.input_physical,positive(P.p90_normalized_hvi),'Color',[.84 .62 .10],'LineWidth',1.6,'DisplayName','90th percentile'); semilogy(ax,P.input_physical,positive(P.p99_normalized_hvi),'Color',[.91 .43 .09],'LineWidth',1.8,'DisplayName','99th percentile'); semilogy(ax,P.input_physical,positive(P.mean_normalized_hvi),'--','Color',[.09 .41 .67],'LineWidth',1.7,'DisplayName','Conditional mean'); semilogy(ax,P.input_physical,positive(P.sampled_profile_maximum_normalized_hvi),'Color',[.13 .13 .13],'LineWidth',2,'DisplayName','Sampled profile maximum'); scatter(ax,P.selected_input_physical(1),P.selected_hvi_normalized(1),72,[.70 .23 .23],'p','filled','Clipping','off','DisplayName','Selected design'); set(ax,'YScale','log'); ylim(ax,[1e-8 1.15]); grid(ax,'on'); xlabel(ax,'Aspect ratio, AR [-]'); ylabel(ax,'Normalized HVI [-]'); title(ax,'B. Conditional quantiles and profile'); legend(ax,'Location','southwest','FontSize',7);
lastPositive=find(P.positive_hvi_fraction>0,1,'last');
if ~isempty(lastPositive) && lastPositive < height(P)
    zeroStart=P.input_physical(lastPositive+1);
    xline(ax,zeroStart,':','Color',[.47 .47 .47],'LineWidth',.8,'HandleVisibility','off');
    text(ax,.98,.20,sprintf('All sampled HVI = 0 for AR >= %g',zeroStart),'Units','normalized', ...
        'HorizontalAlignment','right','FontSize',8,'BackgroundColor','w','Interpreter','none');
end

ax=nexttile(tl); hold(ax,'on'); semilogy(ax,P.input_physical,positive(P.positive_hvi_fraction),'Color',[.91 .43 .09],'LineWidth',2,'DisplayName','Positive-HVI fraction'); semilogy(ax,P.input_physical,positive(P.mean_normalized_hvi),'--','Color',[.09 .41 .67],'LineWidth',1.8,'DisplayName','Conditional mean'); set(ax,'YScale','log'); ylim(ax,[1e-8 1]); grid(ax,'on'); xlabel(ax,'Aspect ratio, AR [-]'); ylabel(ax,'Conditional domain-volume statistic [-]'); title(ax,'C. Positive-HVI support and conditional mean'); legend(ax,'Location','best','FontSize',8);
allAxes=findall(fig,'Type','axes'); for k=1:numel(allAxes), xlim(allAxes(k),[min(x) max(x)]); set(allAxes(k),'YMinorGrid','off','XMinorGrid','off'); end
title(tl,'Frozen sampled-HVI field versus aspect ratio before evaluation 100','FontWeight','bold'); exportgraphics(fig,outputFile,'Resolution',300); close(fig); disp(outputFile);
end
function y=positive(x), y=x; y(y<=0)=NaN; end
